File: training/net.py
"""AlphaZero-style ResNet for chess.

4 residual blocks × 96 filters. Policy head outputs 4672 logits, value
head outputs a scalar in [-1, 1] (tanh) from the side-to-move's POV.
"""

import torch
import torch.nn as nn
import torch.nn.functional as F

from encode import INPUT_CHANNELS, POLICY_SIZE


N_BLOCKS = 4
N_FILTERS = 96


class ResBlock(nn.Module):
    def __init__(self, filters):
        super().__init__()
        self.conv1 = nn.Conv2d(filters, filters, 3, padding=1, bias=False)
        self.bn1 = nn.BatchNorm2d(filters)
        self.conv2 = nn.Conv2d(filters, filters, 3, padding=1, bias=False)
        self.bn2 = nn.BatchNorm2d(filters)

    def forward(self, x):
        h = F.relu(self.bn1(self.conv1(x)))
        h = self.bn2(self.conv2(h))
        return F.relu(h + x)


class ChessNet(nn.Module):
    def __init__(self, n_blocks=N_BLOCKS, n_filters=N_FILTERS):
        super().__init__()
        self.stem_conv = nn.Conv2d(INPUT_CHANNELS, n_filters, 3, padding=1, bias=False)
        self.stem_bn = nn.BatchNorm2d(n_filters)
        self.tower = nn.ModuleList([ResBlock(n_filters) for _ in range(n_blocks)])

        # Policy head: 1×1 conv to 2 channels, then linear to POLICY_SIZE.
        self.policy_conv = nn.Conv2d(n_filters, 2, 1, bias=False)
        self.policy_bn = nn.BatchNorm2d(2)
        self.policy_fc = nn.Linear(2 * 8 * 8, POLICY_SIZE)

        # Value head: 1×1 conv to 1 channel, linear→256→1, tanh.
        self.value_conv = nn.Conv2d(n_filters, 1, 1, bias=False)
        self.value_bn = nn.BatchNorm2d(1)
        self.value_fc1 = nn.Linear(8 * 8, 256)
        self.value_fc2 = nn.Linear(256, 1)

    def forward(self, x):
        h = F.relu(self.stem_bn(self.stem_conv(x)))
        for block in self.tower:
            h = block(h)

        p = F.relu(self.policy_bn(self.policy_conv(h)))
        p = self.policy_fc(p.flatten(1))

        v = F.relu(self.value_bn(self.value_conv(h)))
        v = F.relu(self.value_fc1(v.flatten(1)))
        v = torch.tanh(self.value_fc2(v))

        return p, v.squeeze(-1)


def n_params(net):
    return sum(p.numel() for p in net.parameters())

File: training-reinforce/selfplay_rs/src/net.rs
//! The 4x96 AlphaZero ResNet (training/net.py `ChessNet`) in tch-rs.
//!
//! Explicit-BatchNorm forward only (matches PyTorch bit-exactly) — this worker is
//! not throughput-critical the way the MCTS one is, so we skip the BN-folding /
//! AOTInductor / CUDA-graph machinery of training/selfplay_rs and keep a single
//! parity-exact path. Weights load from fp32 safetensors keyed by state_dict FQN
//! (stem_conv/stem_bn, tower.{i}.{conv1,bn1,conv2,bn2}, policy_conv/policy_bn/
//! policy_fc, value_conv/value_bn/value_fc1/value_fc2).

use std::collections::HashMap;

use tch::{Device, Kind, Tensor};

const BN_EPS: f64 = 1e-5;
const BN_MOMENTUM: f64 = 0.1;

pub struct Net {
    p: HashMap<String, Tensor>,
    n_blocks: usize,
}

fn load_map(path: &str, dev: Device, kind: Kind) -> HashMap<String, Tensor> {
    Tensor::read_safetensors(path)
        .unwrap_or_else(|e| panic!("read_safetensors {path}: {e}"))
        .into_iter()
        .map(|(k, v)| {
            // BN tracks an int64 num_batches_tracked buffer; keep its dtype.
            let t = if k.ends_with("num_batches_tracked") {
                v.to_device(dev)
            } else {
                v.to_device(dev).to_kind(kind)
            };
            (k, t)
        })
        .collect()
}

impl Net {
    pub fn load(path: &str, dev: Device, kind: Kind) -> Net {
        let p = load_map(path, dev, kind);
        let n_blocks = (0..)
            .take_while(|i| p.contains_key(&format!("tower.{i}.conv1.weight")))
            .count();
        Net { p, n_blocks }
    }

    fn g(&self, k: &str) -> &Tensor {
        self.p.get(k).unwrap_or_else(|| panic!("missing param {k}"))
    }

    fn conv(&self, x: &Tensor, w: &str, pad: i64) -> Tensor {
        let none: Option<Tensor> = None;
        x.conv2d(self.g(w), none.as_ref(), [1, 1], [pad, pad], [1, 1], 1)
    }

    fn bn(&self, h: &Tensor, pfx: &str) -> Tensor {
        h.batch_norm(
            Some(self.g(&format!("{pfx}.weight"))),
            Some(self.g(&format!("{pfx}.bias"))),
            Some(self.g(&format!("{pfx}.running_mean"))),
            Some(self.g(&format!("{pfx}.running_var"))),
            false, // not training: use running stats
            BN_MOMENTUM,
            BN_EPS,
            true, // cudnn enabled
        )
    }

    /// (policy_logits [B,4672], value [B]) — matches ChessNet.forward exactly.
    pub fn forward(&self, x: &Tensor) -> (Tensor, Tensor) {
        let mut h = self.bn(&self.conv(x, "stem_conv.weight", 1), "stem_bn").relu();
        for i in 0..self.n_blocks {
            let inp = h.shallow_clone();
            let k = self
                .bn(&self.conv(&inp, &format!("tower.{i}.conv1.weight"), 1), &format!("tower.{i}.bn1"))
                .relu();
            let k = self.bn(&self.conv(&k, &format!("tower.{i}.conv2.weight"), 1), &format!("tower.{i}.bn2"));
            h = (k + inp).relu();
        }

        let policy = self
            .bn(&self.conv(&h, "policy_conv.weight", 0), "policy_bn")
            .relu()
            .flatten(1, -1)
            .linear(self.g("policy_fc.weight"), Some(self.g("policy_fc.bias")));

        let value = self
            .bn(&self.conv(&h, "value_conv.weight", 0), "value_bn")
            .relu()
            .flatten(1, -1)
            .linear(self.g("value_fc1.weight"), Some(self.g("value_fc1.bias")))
            .relu()
            .linear(self.g("value_fc2.weight"), Some(self.g("value_fc2.bias")))
            .tanh()
            .squeeze_dim(-1);

        (policy, value)
    }
}
